File: ios/Classes/WhisperPlugin.swift
import Flutter
import whisper
import Foundation
import AVFoundation

// Declare C bridge functions from WhisperBridge.mm
@_silgen_name("whisper_init_bridge")
func whisper_init_bridge(_ path: UnsafePointer<CChar>) -> OpaquePointer?

@_silgen_name("whisper_full_bridge")
func whisper_full_bridge(_ ctx: OpaquePointer?, _ samples: UnsafeMutablePointer<Float>?, _ n: Int32) -> Int32

@_silgen_name("whisper_free_bridge")
func whisper_free_bridge(_ ctx: OpaquePointer?)

@_silgen_name("whisper_get_result_bridge")
func whisper_get_result_bridge(_ ctx: OpaquePointer?) -> UnsafePointer<CChar>?

@_silgen_name("whisper_free_cstr_bridge")
func whisper_free_cstr_bridge(_ s: UnsafePointer<CChar>?)

@objc(WhisperPlugin)
public class WhisperPlugin: NSObject, FlutterPlugin {
    private var cachedContext: OpaquePointer? = nil
    private var cachedModelPath: String? = nil
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        // 👇 Empêche le linker de supprimer whisper.xcframework
        EnforceBinding.dummyMethodToEnforceBundling()
        
        let channel = FlutterMethodChannel(name: "flutter_whisper_ggml", binaryMessenger: registrar.messenger())
        let instance = WhisperPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "transcribe":
            guard let args = call.arguments as? [String: Any],
                  let filePath = args["filePath"] as? String,
                  let modelPath = args["modelPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "filePath and modelPath are required", details: nil))
                return
            }
            
            transcribeAsync(filePath: filePath, modelPath: modelPath, result: result)
        case "transcribeData":
            guard let args = call.arguments as? [String: Any],
                  let modelPath = args["modelPath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "modelPath and audioData are required", details: nil))
                return
            }
            
            if let doubles = args["audioData"] as? [Double] {
                let floatSamples = doubles.map { Float($0) }
                transcribeBuffer(modelPath: modelPath, samples: floatSamples, result: result)
            } else if let typed = args["audioData"] as? FlutterStandardTypedData {
                let byteCount = typed.data.count
                if byteCount % MemoryLayout<Float>.size != 0 {
                    result(FlutterError(code: "INVALID_ARGUMENT", message: "audioData has invalid byte length", details: nil))
                    return
                }
                let count = byteCount / MemoryLayout<Float>.size
                var floatSamples = [Float](repeating: 0, count: count)
                _ = floatSamples.withUnsafeMutableBytes { mutablePointer in
                    typed.data.copyBytes(to: mutablePointer)
                }
                transcribeBuffer(modelPath: modelPath, samples: floatSamples, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Unsupported audioData format", details: nil))
            }
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func transcribeAsync(filePath: String, modelPath: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "ERROR", message: "Plugin deallocated", details: nil))
                }
                return
            }
            
            do {
                // Load or reuse context
                let context = try self.getOrCreateContext(modelPath: modelPath)
                
                // Read WAV file
                let audioData = try self.readWavFile(filePath: filePath)
                
                // Transcribe
                let transcript = try self.transcribe(context: context, audioData: audioData)
                
                DispatchQueue.main.async {
                    result(transcript)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "TRANSCRIPTION_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func transcribeBuffer(modelPath: String, samples: [Float], result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "ERROR", message: "Plugin deallocated", details: nil))
                }
                return
            }
            
            do {
                let context = try self.getOrCreateContext(modelPath: modelPath)
                let transcript = try self.transcribe(context: context, audioData: samples)
                
                DispatchQueue.main.async {
                    result(transcript)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "TRANSCRIPTION_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func getOrCreateContext(modelPath: String) throws -> OpaquePointer {
        // Reuse cached context if same model path
        if let cached = cachedContext, cachedModelPath == modelPath {
            return cached
        }
        
        // Release old context if model path changed
        if let oldContext = cachedContext {
            whisper_free_bridge(oldContext)
            cachedContext = nil
            cachedModelPath = nil
        }
        
        // Load new model
        guard let context = whisper_init_bridge(modelPath) else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to load Whisper model"])
        }
        
        cachedContext = context
        cachedModelPath = modelPath
        
        return context
    }
    
    private func readWavFile(filePath: String) throws -> [Float] {
        let url = URL(fileURLWithPath: filePath)
        let data = try Data(contentsOf: url)
        
        guard data.count >= 44 else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "WAV file too small"])
        }
        
        // Check RIFF header
        let riff = String(data: data[0..<4], encoding: .ascii)
        guard riff == "RIFF" else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid WAV file: missing RIFF header"])
        }
        
        // Check WAVE header
        let wave = String(data: data[8..<12], encoding: .ascii)
        guard wave == "WAVE" else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid WAV file: missing WAVE header"])
        }
        
        // Parse chunks
        var pos = 12
        var sampleRate: Int? = nil
        var numChannels: Int? = nil
        var bitsPerSample: Int? = nil
        var dataStart: Int? = nil
        var dataSize: Int? = nil
        
        while pos < data.count - 8 {
            let chunkId = String(data: data[pos..<pos+4], encoding: .ascii) ?? ""
            let chunkSize = data.withUnsafeBytes { bytes in
                bytes.load(fromByteOffset: pos + 4, as: UInt32.self).littleEndian
            }
            
            if chunkId == "fmt " {
                // Parse fmt chunk
                let audioFormat = data.withUnsafeBytes { bytes in
                    Int(bytes.load(fromByteOffset: pos + 8, as: UInt16.self).littleEndian)
                }
                numChannels = data.withUnsafeBytes { bytes in
                    Int(bytes.load(fromByteOffset: pos + 10, as: UInt16.self).littleEndian)
                }
                sampleRate = data.withUnsafeBytes { bytes in
                    Int(bytes.load(fromByteOffset: pos + 12, as: UInt32.self).littleEndian)
                }
                bitsPerSample = data.withUnsafeBytes { bytes in
                    Int(bytes.load(fromByteOffset: pos + 22, as: UInt16.self).littleEndian)
                }
                
                if audioFormat != 1 {
                    throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unsupported audio format: \(audioFormat) (only PCM supported)"])
                }
            } else if chunkId == "data" {
                dataStart = pos + 8
                dataSize = Int(chunkSize)
                break
            }
            
            pos += 8 + Int(chunkSize)
            if chunkSize % 2 == 1 { pos += 1 }
        }
        
        guard let sr = sampleRate, let ch = numChannels, let bps = bitsPerSample,
              let ds = dataStart, let dsz = dataSize else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Missing fmt or data chunk in WAV file"])
        }
        
        // Validate format
        if ch != 1 || sr != 16000 || bps != 16 {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "WAV must be PCM16 mono @16kHz"])
        }
        
        // Determine effective data size: some recorders may leave data chunk size to 0 until finalization
        var effectiveDataSize = dsz
        if dsz == 0 || ds + dsz > data.count {
            effectiveDataSize = max(0, data.count - ds)
            NSLog("⚠️ [WHISPER] WAV data chunk reported size %d, using effective size %d", dsz, effectiveDataSize)
        }
        
        // Read PCM16 samples
        let bytesPerSample = bps / 8
        let totalSamples = effectiveDataSize / bytesPerSample / ch
        if totalSamples <= 0 {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "No audio samples in WAV data chunk"])
        }
        
        // Convert to float32
        var floatArray = [Float](repeating: 0, count: totalSamples)
        for i in 0..<totalSamples {
            let sampleOffset = ds + i * bytesPerSample * ch
            let sampleInt = data.withUnsafeBytes { bytes in
                Int16(bitPattern: UInt16(bytes.load(fromByteOffset: sampleOffset, as: UInt16.self).littleEndian))
            }
            floatArray[i] = Float(sampleInt) / 32768.0
        }
        
        return floatArray
    }
    
    private func transcribe(context: OpaquePointer, audioData: [Float]) throws -> String {
        // Use default Whisper parameters (like the fast example)
        let result = audioData.withUnsafeBufferPointer { buffer in
            whisper_full_bridge(context, UnsafeMutablePointer(mutating: buffer.baseAddress), Int32(buffer.count))
        }
        
        guard result == 0 else {
            let msg = "Whisper transcription failed (code \(result))"
            NSLog("❌ [WHISPER] %@", msg)
            throw NSError(domain: "WhisperPlugin", code: Int(result), userInfo: [NSLocalizedDescriptionKey: msg])
        }
        
        // Get result
        guard let resultPtr = whisper_get_result_bridge(context) else {
            throw NSError(domain: "WhisperPlugin", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to get transcription result"])
        }
        
        let transcript = String(cString: resultPtr)
        whisper_free_cstr_bridge(resultPtr)
        
        return transcript
    }
    
    deinit {
        if let context = cachedContext {
            whisper_free_bridge(context)
        }
    }
}

File: android/src/main/kotlin/fr/yawendaba/flutter_whisper_ggml/WhisperPlugin.kt
package fr.yawendaba.flutter_whisper_ggml

import android.util.Log
import com.whispercpp.whisper.WhisperContext
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

class WhisperPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_whisper_ggml")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "transcribe" -> {
                Log.d("WhisperPlugin", "transcribe called")
                val filePath = call.argument<String>("filePath")
                val modelPath = call.argument<String>("modelPath")

                if (filePath == null || modelPath == null) {
                    result.error("INVALID_ARGUMENT", "filePath and modelPath are required", null)
                    return
                }

                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val floatArray = readWavFile(filePath)
                        val whisper = WhisperContext.createContextFromFile(modelPath)
                        val text = whisper.transcribeData(floatArray, printTimestamp = false)
                        whisper.release()
                        
                        withContext(Dispatchers.Main) {
                            result.success(text)
                        }
                    } catch (e: Exception) {
                        Log.e("WhisperPlugin", "Error during transcription", e)
                        withContext(Dispatchers.Main) {
                            result.error("TRANSCRIPTION_ERROR", e.message, null)
                        }
                    }
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun readWavFile(filePath: String): FloatArray {
        val file = File(filePath)
        val bytes = file.readBytes()

        // Parse WAV header
        if (bytes.size < 44) {
            throw IllegalArgumentException("WAV file too small")
        }

        // Check RIFF header
        val riff = String(bytes.sliceArray(0..3))
        if (riff != "RIFF") {
            throw IllegalArgumentException("Invalid WAV file: missing RIFF header")
        }

        // Check WAVE header
        val wave = String(bytes.sliceArray(8..11))
        if (wave != "WAVE") {
            throw IllegalArgumentException("Invalid WAV file: missing WAVE header")
        }

        // Parse chunks
        var pos = 12
        var sampleRate: Int? = null
        var numChannels: Int? = null
        var bitsPerSample: Int? = null
        var dataStart: Int? = null
        var dataSize: Int? = null

        while (pos < bytes.size - 8) {
            val chunkId = String(bytes.sliceArray(pos until pos + 4))
            val chunkSize = ByteBuffer.wrap(bytes, pos + 4, 4)
                .order(ByteOrder.LITTLE_ENDIAN)
                .int

            if (chunkId == "fmt ") {
                // Parse fmt chunk
                val audioFormat = ByteBuffer.wrap(bytes, pos + 8, 2)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .short.toInt()
                numChannels = ByteBuffer.wrap(bytes, pos + 10, 2)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .short.toInt()
                sampleRate = ByteBuffer.wrap(bytes, pos + 12, 4)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .int
                bitsPerSample = ByteBuffer.wrap(bytes, pos + 22, 2)
                    .order(ByteOrder.LITTLE_ENDIAN)
                    .short.toInt()

                if (audioFormat != 1) {
                    throw IllegalArgumentException("Unsupported audio format: $audioFormat (only PCM supported)")
                }
            } else if (chunkId == "data") {
                dataStart = pos + 8
                dataSize = chunkSize
                break
            }

            pos += 8 + chunkSize
            if (chunkSize % 2 == 1) pos++
        }

        if (sampleRate == null || numChannels == null || bitsPerSample == null) {
            throw IllegalArgumentException("Missing fmt chunk in WAV file")
        }

        if (dataStart == null || dataSize == null) {
            throw IllegalArgumentException("Missing data chunk in WAV file")
        }

        // Validate format
        if (numChannels != 1 || sampleRate != 16000 || bitsPerSample != 16) {
            throw IllegalArgumentException("WAV must be PCM16 mono @16kHz")
        }

        // Read PCM16 samples
        val bytesPerSample = bitsPerSample / 8
        val totalSamples = dataSize / bytesPerSample / numChannels

        // Convert to float32
        val floatArray = FloatArray(totalSamples)
        for (i in 0 until totalSamples) {
            val sampleOffset = dataStart + i * bytesPerSample * numChannels
            val sampleInt = ByteBuffer.wrap(bytes, sampleOffset, 2)
                .order(ByteOrder.LITTLE_ENDIAN)
                .short.toInt()
            floatArray[i] = sampleInt / 32768.0f
        }

        // Trim silence
        val trimmed = trimSilence(floatArray)
        Log.d("WhisperPlugin", "Audio samples: ${floatArray.size} -> ${trimmed.size} (after trim)")

        return trimmed
    }

    private fun trimSilence(samples: FloatArray, threshold: Float = 0.02f): FloatArray {
        var i0 = 0
        var i1 = samples.size

        while (i0 < i1 && kotlin.math.abs(samples[i0]) < threshold) i0++
        while (i1 > i0 && kotlin.math.abs(samples[i1 - 1]) < threshold) i1--

        return if (i0 > 0 || i1 < samples.size) {
            samples.sliceArray(i0 until i1)
        } else {
            samples
        }
    }
}

